xla/hlo/analysis/while_loop_analysis.cc

```diff
@@ -354,6 +355,29 @@ optional<int64_t> CheckedSubtract(int64_t a, int64_t b) {
   return result;
 }
 
+// This function returns true if the operation is a simple scalar operation.
+// While loop analysis can execute such an operation at compile time without
+// incurring huge overheads.
+bool IsScalarOp(const HloInstruction* op) {
+  if (IsCollective(op)) return false;
+  switch (op->opcode()) {
+    case HloOpcode::kSend:
+    case HloOpcode::kSendDone:
+    case HloOpcode::kRecv:
+    case HloOpcode::kRecvDone:
+    case HloOpcode::kCustomCall:
+      return false;
+    default:
+      break;
+  }
```

Comment:
Can you please also check whether op->shape() is a scalar (ShapeUtil::IsEffectiveScalar())? If each op in the while condition computation has a scalar shape, we can be relatively sure that it can be evaluated quickly. Only exception: parameters of the computation may have larger shapes.

Reply:
Done.